creme/linear_model/adpredictor.py

```diff
@@ -0,0 +1,175 @@
+        surprise (float): Clipping value of the learning step function parameter ensure numerical stability.
+        epsilon (float): Noise parameter of the dynamic correction process.
+
+    Attributes:
+        weights (collections.defaultdict): The current gaussian belief over the weights.
+        
+    """
+    def __init__(self, initializer, surprise, epsilon):
+        self.surprise = surprise
+        self.epsilon = epsilon
+        self.beta = initializer.beta
+        self.prior_probability = initializer.prior_probability
+        self.initializer = initializer
+        self.weights = collections.defaultdict(initializer)
+
+    def _active_mean_variance(self, x):
+        means = [self.weights[i].mean for i, xi in x.items()]
+        variances = [self.weights[i].variance for i, xi in x.items()]
+
+        return sum(means), sum(variances) + self.beta ** 2
+
+    def _gaussian_correction(self, t):
+        """TODO 
+        """
+        t = utils.math.clamp(t, minimum=-self.surprise, maximum=self.surprise)
+        v = stats.norm.pdf(t)/stats.norm.cdf(t)
+        w = v * (v + t)
+        return (v, w)
+
+    def _apply_dynamics(self, weight):
+        """TODO
+        """
+        prior = self.initializer
+        adjusted_variance = weight.variance * prior.variance / ((1.0 - self.epsilon) * prior.variance + self.epsilon * weight.variance)
+        adjusted_mean = adjusted_variance * ((1.0 - self.epsilon) * weight.mean / weight.variance + self.epsilon * prior.mean / prior.variance)
+        prior.variance = adjusted_variance
+        prior.mean = adjusted_mean
+        return prior
+
+    @staticmethod
+    def _target_encoding(y):
+        assert isinstance(y, numbers.Number)
+        return 1.0 if y == 1 else -1.0
+
+
+class AdPredictor(BOPR, base.BinaryClassifier):
+    """AdPredictor is a bayesian online probit regression model devised for search advertising click through rate (CTR) prediction.
+
+    AdPredictor
+
+    Parameters:
+        initializer (linear_model.adpredictor.NormalPrior): The initializer used to set beliefs over the weights.
+        surprise (float): Clipping value of the learning step function parameter ensure numerical stability.
+        epsilon (float): Noise parameter of the dynamic correction process.
+
+    Attributes:
+        weights (collections.defaultdict): The current gaussian belief over the weights.
+
+    Example:
+
+        ::
+
+            >>> X_y = datasets.Phishing()
+
+            >>> model = (
+            ...     compose.Select('age_of_domain','ip_in_url')|
+            ...     preprocessing.OneHotEncoder('age_of_domain')|
+            ...     linear_model.AdPredictor(initializer=NormalPrior(beta=0.1, n_features=X_y.n_features, prior_probability=0.6), surprise=5, epsilon=0.001)
+            ... )
+
+            >>> metric = metrics.ROCAUC()
+
+            >>> model_selection.progressive_val_score(X_y, model, metric)
+            ROCAUC: 0.540405
```

Comment:
That doesn't look too good! It might be nice to take Andrew Tulloch's example and to check whether you're able to reach the same performance as he does.

Reply:
Ahahah yes it's pretty bad I  benchmarked it against the logistic regression on same features `ROCAUC: 0.617649`. I will refactor the code and see if I didn't miss anything and benchmark it with Andrew Tulloch's implementation  !